cjwstate/modules/message_catalogs.py

```diff
@@ -0,0 +1,116 @@
+from cjworkbench.i18n import supported_locales, default_locale
+from cjworkbench.i18n.catalogs.util import (
+    read_po_catalog,
+    write_po_catalog,
+    remove_strings,
+    find_fuzzy_messages,
+    fill_catalog,
+    mark_fuzzy,
+    MessageUID,
+)
+from babel.messages.catalog import Catalog
+from cjwstate.modules.module_loader import ModuleFiles, ModuleSpec
+import os.path
+from typing import FrozenSet
+
+
+def update_module_catalogs(spec: ModuleSpec):
+    source_catalog = _build_source_catalog(spec)
+    po_path = _po_path(directory, default_locale, spec.id_name)
+    try:
+        old_source_catalog = read_po_catalog(po_path)
+    except FileNotFoundError:
+        old_source_catalog = Catalog(default_locale)
+    fuzzy = find_fuzzy_messages(
+        old_catalog=old_source_catalog, new_catalog=source_catalog
+    )
+    write_po_catalog(po_path, source_catalog)
+
+    for locale_id in supported_locales:
+        po_path = _po_path(directory, locale_id, spec.id_name)
+        try:
+            old_catalog = read_po_catalog(po_path)
+        except FileNotFoundError:
+            old_catalog = Catalog(locale_id)
+        write_po_catalog(
+            po_path, _merge_nonsource_catalog(old_catalog, source_catalog, fuzzy)
+        )
+
+
+def _po_path(basepath: os.path, locale_id: str, module_id_name: str) -> os.path:
+    return os.path.join(basepath, "locale", locale_id, f"{module_id_name}.po")
+
+
+def _build_source_catalog(spec: ModuleSpec) -> Catalog:
+    messages = _find_spec_messages(spec)
+    # TODO: also find messages in module code
+    source_catalog = Catalog(default_locale)
+    for message_id, source_string in messages.items():
+        if source_string is not None:
+            source_catalog.add(message_id, string=source_string)
+
+    return source_catalog
+
+
+def _merge_nonsource_catalog(
+    old_catalog: Catalog, source_catalog: Catalog, fuzzy: FrozenSet[MessageUID]
+) -> Catalog:
+    catalog = Catalog(locale_id)
+    fill_catalog(catalog, source_catalog, old_catalog)
+    mark_fuzzy(catalog, fuzzy, old_catalog)
+    return catalog
+
+
+def _find_spec_messages(spec: ModuleSpec) -> Catalog:
+    prefix = f"modules.{spec.id_name}._spec"
+    messages = {
+        f"{prefix}.name": spec.get("name"),
+        f"{prefix}.description": spec.get("description"),
+        f"{prefix}.row_action_menu_entry_title": spec.get(
+            "row_action_menu_entry_title"
+        ),
+        f"{prefix}.deprecated.message": spec.get("deprecated", {}).get("message"),
+    }
+    for param in spec.parameters:
+        messages = {
+            **_find_param_messages(param, f"{prefix}.parameters.{param['id_name']}"),
+            **messages,
+        }
+    return messages
+
+
+def _find_param_messages(param, param_prefix):
```

Comment:
First off: please declare types here. I was expecting a `ParamSpec`, but this looks like a `Dict`?

Second: the correct way to visit trees is the visitor design pattern. We use Python `@singledispatch` for this. Look to `renderprep.py` for an example. (renderprep uses ParamDType, but the same pattern works for ParamSpec.) Every type is listed in `cjwstate/models/param_spec.py`. Keep in mind that optional-string fields usually have empty strings: we don't want to translate empty strings. (We assume the module author designed empty string for a reason.)

So something like:

```
def _add_if_set(l: List[Tuple[str, str]], spec: ParamSpec, prefix: str, key: str) -> None:
    value = spec.getattr(key)
    if value:
        l.append((f"{prefix}.{key}", value)


@singledispatch
def extract_param_messages(spec: ParamSpec, prefix: str) -> List[Tuple[str, str]]:
    return []  # fallback for types we don't handle


@extract_param_messages.register(ParamSpec.String)
def _(spec: ParamSpec.String, prefix: str) -> List[Tuple[str, str]]:
    ret = []
    _add_if_set(ret, spec, prefix, "name")
    _add_if_set(ret, spec, prefix, "placeholder")
    _add_if_set(ret, spec, prefix, "default")
    return ret

@extract_param_messages.register(ParamSpec.Secret)
def _(...)
```

... this ends up being rather verbose, so I recommend putting it in a separate Python module.

The win is: correctness. Treat `param_spec.py` as a to-do list: it lists all types and all fields for those types.